### PR TITLE
Gdamore/win io lock

### PR DESCRIPTION
**Pull request by gdamore**
This is a trial run, and it may or may not work.

### CODE REVIEW

File: src/platform/windows/win_io.c [hunk 1]
```diff
@@ -23,6 +23,8 @@ static int      win_io_nthr = 0;
 static HANDLE   win_io_h    = NULL;
 static nni_thr *win_io_thrs;
 
+static SRWLock win_io_lock = SRWLOCK_INIT;
```
Comment (alzix) [inline src/platform/windows/win_io.c:26]:
```suggestion
static SRWLOCK win_io_lock = SRWLOCK_INIT;
```

File: src/platform/windows/win_io.c [hunk 2]
```diff
@@ -61,26 +69,44 @@ nni_win_io_register(HANDLE h)
...
 
 void
 nni_win_io_fini(nni_win_io *io)
 {
-	if (io->olpd.hEvent != NULL) {
-		CloseHandle((HANDLE) io->olpd.hEvent);
+	DWORD num;
+	// Make absolutely sure there is no I/O running.
+	if (io->f != INVALID_HANDLE) {
+		CancelIoEx(io->f, &o->olpd);
```
Comment (alzix) [inline src/platform/windows/win_io.c:89]:
```suggestion
		CancelIoEx(io->f, &io->olpd);
```

File: src/platform/windows/win_io.c [hunk 3]
```diff
@@ -61,26 +69,44 @@ nni_win_io_register(HANDLE h)
...
 }
 
 void
 nni_win_io_fini(nni_win_io *io)
 {
-	if (io->olpd.hEvent != NULL) {
-		CloseHandle((HANDLE) io->olpd.hEvent);
+	DWORD num;
+	// Make absolutely sure there is no I/O running.
+	if (io->f != INVALID_HANDLE) {
```
Comment (alzix) [inline src/platform/windows/win_io.c:88]:
```suggestion
	if (io->f != INVALID_HANDLE_VALUE) {
```

File: src/platform/windows/win_ipclisten.c [hunk 4]
```diff
@@ -331,7 +331,8 @@ nni_ipc_listener_alloc(nng_stream_listener **lp, const nng_url *url)
 	if ((l = NNI_ALLOC_STRUCT(l)) == NULL) {
 		return (NNG_ENOMEM);
 	}
-	if ((rv = nni_win_io_init(&l->io, ipc_accept_cb, l)) != 0) {
+	if ((rv = nni_win_io_init(&l->io, INVALID_HANDLE, ipc_accept_cb, l)) !=
```
Comment (alzix) [inline src/platform/windows/win_ipclisten.c:334]:
```suggestion
	if ((rv = nni_win_io_init(&l->io, INVALID_HANDLE_VALUE, ipc_accept_cb, l)) !=
```

File: src/platform/windows/win_io.c [hunk 5]
```diff
@@ -61,26 +69,44 @@ nni_win_io_register(HANDLE h)
 }
 
 int
-nni_win_io_init(nni_win_io *io, nni_win_io_cb cb, void *ptr)
+nni_win_io_init(nni_win_io *io, HANDLE h, nni_win_io_cb cb, void *ptr)
```
Comment (alzix) [inline src/platform/windows/win_io.c:72]: need to update the header file as well

File: src/platform/windows/win_tcpconn.c [hunk 6]
```diff
@@ -448,8 +448,8 @@ nni_win_tcp_init(nni_tcp_conn **connp, SOCKET s)
 	c->ops.s_get   = tcp_get;
 	c->ops.s_set   = tcp_set;
 
-	if (((rv = nni_win_io_init(&c->recv_io, tcp_recv_cb, c)) != 0) ||
-	    ((rv = nni_win_io_init(&c->send_io, tcp_send_cb, c)) != 0) ||
+	if (((rv = nni_win_io_init(&c->recv_io, s, tcp_recv_cb, c)) != 0) ||
+	    ((rv = nni_win_io_init(&c->send_io, s, tcp_send_cb, c)) != 0) ||
```
Comment (alzix) [inline src/platform/windows/win_tcpconn.c:452]:
```suggestion
	if (((rv = nni_win_io_init(&c->recv_io, (HANDLE) s, tcp_recv_cb, c)) != 0) ||
	    ((rv = nni_win_io_init(&c->send_io, (HANDLE) s, tcp_send_cb, c)) != 0) ||
```

File: src/platform/windows/win_tcpconn.c [hunk 7]
```diff
@@ -448,8 +448,8 @@ nni_win_tcp_init(nni_tcp_conn **connp, SOCKET s)
 	c->ops.s_get   = tcp_get;
 	c->ops.s_set   = tcp_set;
 
-	if (((rv = nni_win_io_init(&c->recv_io, tcp_recv_cb, c)) != 0) ||
-	    ((rv = nni_win_io_init(&c->send_io, tcp_send_cb, c)) != 0) ||
+	if (((rv = nni_win_io_init(&c->recv_io, s, tcp_recv_cb, c)) != 0) ||
```
Comment (alzix) [inline src/platform/windows/win_tcpconn.c:451]: please remember to update the function declaration in `win_impl.h:125`

File: src/platform/windows/win_tcplisten.c [hunk 8]
```diff
@@ -322,7 +322,7 @@ nni_tcp_listener_accept(nni_tcp_listener *l, nni_aio *aio)
 	c->listener = l;
 	c->conn_aio = aio;
 	nni_aio_set_prov_data(aio, c);
-	if (((rv = nni_win_io_init(&c->conn_io, tcp_accept_cb, c)) != 0) ||
+	if (((rv = nni_win_io_init(&c->conn_io, s, tcp_accept_cb, c)) != 0) ||
```
Comment (alzix) [inline src/platform/windows/win_tcplisten.c:325]:
```suggestion
	if (((rv = nni_win_io_init(&c->conn_io, (HANDLE) s, tcp_accept_cb, c)) != 0) ||
```

File: src/platform/windows/win_tcpdial.c [hunk 9]
```diff
@@ -208,7 +208,7 @@ nni_tcp_dial(nni_tcp_dialer *d, const nni_sockaddr *sa, nni_aio *aio)
 
 	c->peername = ss;
 
-	if ((rv = nni_win_io_init(&c->conn_io, tcp_dial_cb, c)) != 0) {
+	if ((rv = nni_win_io_init(&c->conn_io, s, tcp_dial_cb, c)) != 0) {
```
Comment (alzix) [inline src/platform/windows/win_tcpdial.c:211]:
```suggestion
	if ((rv = nni_win_io_init(&c->conn_io, (HANDLE) s, tcp_dial_cb, c)) != 0) {
```

File: src/platform/windows/win_udp.c [hunk 10]
```diff
@@ -67,7 +67,7 @@ nni_plat_udp_open(nni_plat_udp **udpp, nni_sockaddr *sa)
 	(void) setsockopt(
 	    u->s, IPPROTO_IPV6, IPV6_V6ONLY, (char *) &no, sizeof(no));
 
-	if (((rv = nni_win_io_init(&u->rxio, udp_recv_cb, u)) != 0) ||
+	if (((rv = nni_win_io_init(&u->rxio, u->s, udp_recv_cb, u)) != 0) ||
```
Comment (alzix) [inline src/platform/windows/win_udp.c:70]:
```suggestion
	if (((rv = nni_win_io_init(&u->rxio, (HANDLE) u->s, udp_recv_cb, u)) != 0) ||
```

**Review by alzix — COMMENTED** [review]
the code does compile after applying the suggestions - but the use-after-free is not solved

**Comment by gdamore** [comment]
Hi @alzix - obviously I didn't get a chance to try building this on Windows myself (except via CI/CID) before I ran out of time.  I'll fix these .. .but if you've already done so, can you tell me if the approach at least resolves the problems you've been observing?

**Comment by alzix** [comment]
no i still got a crash 
image
image

**Comment by alzix** [comment]
I also saw the crash in the `nni_stat_inc` especially when running with a debugger and stopping threads for some time.
IMO we better concentrate on this critical issue first. 
Statistics can be compiled out if it cause problems. @KnappSas - can you please open a different issue for it - so it won't get forgotten

**Comment by itayzafrir** [comment]
Hi @gdamore and all.

I've been looking into this and it seems as if the pipe is being deallocated/freed (in `ipc_pipe_fini`) **but** after that a call to `ipc_send_cb` is received (from `win_io_handler`). 
Inside `ipc_send_cb` the `nni_aio` pointer has the same address of the `tx_aio` member of `ipc_pipe`. Since `tx_aio` is a _by value member_ and since the the `ipc_pipe` has been deallocated then this pointer is a dangling pointer.

image

**Comment by gdamore** [comment]
THanks -- I've been looking at this in spare cycles, but very few such cycles exist, which is why it's taking time.  I'll try to get to it in a day or so.

**Comment by gdamore** [comment]
So the design is meant to ensure that all callbacks are done, and the pipe is totally unregistered before we freed it.  I've missed something.  I'm going to resolve this one way the other this weekend.

**Comment by gdamore** [comment]
CLosing I merged a different branch with different approach.

**Pull request closed without merging**